src/app/tap.rs

```diff
@@ -0,0 +1,99 @@
+use super::identity;
+use futures::{future, Future};
+use logging;
+use proxy;
+use std::error;
+use svc;
+use tokio::executor;
+use tower_grpc as grpc;
+use transport::{
+    tls::{self, HasPeerIdentity},
+    Listen,
+};
+use Conditional;
+
+macro_rules! tap_server {
+    { $session:ident, $svc:ident, $log_context:ident } => {{
+        let svc = proxy::grpc::req_box_body::Service::new($svc);
+        let svc = proxy::grpc::res_body_as_payload::Service::new(svc);
+        let svc = proxy::http::HyperServerSvc::new(svc);
+
+        hyper::server::conn::Http::new()
+            .with_executor($log_context.executor())
+            .http2_only(true)
+            .serve_connection($session, svc)
+            .map_err(|err| debug!("tap connection error: {}", err))
+    };}
+}
+
+macro_rules! tap_task {
+    { $srv:ident, $log:ident, $new_service:ident } => {{
+        executor::current_thread::TaskExecutor::current()
+            .spawn_local(Box::new($log.future($srv)))
+            .map(|()| $new_service)
+            .map_err(task::Error::into_io)
+    };}
+}
+
+pub fn serve_tap<N, B>(
+    bound_port: Listen<identity::Local, ()>,
+    tap_svc_name: tls::PeerIdentity,
+    new_service: N,
+) -> impl Future<Item = (), Error = ()> + 'static
+where
+    B: tower_grpc::Body + Send + 'static,
+    B::Data: Send + 'static,
+    N: svc::MakeService<(), http::Request<grpc::BoxBody>, Response = http::Response<B>>
+        + Send
+        + 'static,
+    N::Error: Into<Box<dyn error::Error + Send + Sync>>,
+    N::MakeError: ::std::fmt::Display,
+    <N::Service as svc::Service<http::Request<grpc::BoxBody>>>::Future: Send + 'static,
+{
+    let log = logging::admin().server("tap", bound_port.local_addr());
+
+    let fut = {
+        let log = log.clone();
+        bound_port
+            .listen_and_fold(new_service, move |mut new_service, (session, remote)| {
+                let log = log.clone().with_remote(remote);
+                let log_context = log.clone();
+
+                if let Conditional::Some(ref tap_svc_name) = tap_svc_name {
```

Comment:
Nit/TIOLI: Throughout this function, you use `..._name` to refer to the tap service's identity. I think it would be more descriptive to use variable names like `..._identity` or `..._id` instead.

Reply:
We go back and forth on `..._name` and `..._id`/`..._identity` a decent amount. I ultimately went with `..._name` since the environment variable is `..._NAME`. The fact that we call it a proxy's `..._LOCAL_NAME` and such,  I decided to continue that naming (heh). If you feel any stronger than a nit let me know, but I will otherwise plan to keep it!

Reply:
I don't really care. IMO naming it `id` or `identity` makes it much clearer to someone looking at this file, and I don't think that being the same as the name of the env var is as important as a reader understanding what the variable _is_ without having to follow it back up the call tree. But, this is just a nit, so call it whatever you prefer.

Reply:
Okay I'm going to stick with `..._name` here. I could flip back and forth just as easy, but the fact that the expected value is from an environment variable called `..._SVC_NAME` is enough for me to keep as is.